# Midterm Take Home F1

Student Name: Henan MU

Professor is curious whether he has the same probability to meet a F1 driver if he goes to one of seven shows for the whole duration of the show and if **7 F1 racers go to 7 shows where each racer goes to one and only one show for 7 minutes**, or if **one F1 driver goes to one or more of 7 shows for exactly 7 times and for 7 minutes each time** (so for example, the sole F1 driver could theoretically go to one show only, then leave, then come back, and repeat this seven times for a visit total of 49 minutes without going to any other show). Which of the two experiments give professor a higher probability of meeting a F1 driver?

Can you help professor solve this puzzle in two different ways: a) By using theory and math or Monte Carlo simulations and b) By using combinatorics where you build the sample space for both experiments, then run our p function p(not_alone, sample_space) where not_alone() is a predicate function that describes the "not alone" situation.



## Experiment One: 7 F1 racers go to 7 shows where each racer goes to one and only one show for 7 minutes

### a) By using theory and math or Monte Carlo simulations

Professor goes to only one of seven shows for the whole duration of the show. For the first racer, the probability of they meeting each other is 1/7:

$$P(meet \space first \space racer) = \frac{1}{7}$$
$$P(not \space meet \space first \space racer) = 1 - P(meet \space first \space racer) =1 - \frac{1}{7} =  \frac{6}{7}$$

For other racers, the probability of professor not meeting them is same like $P(not \space meet \space first \space racer)$ which is 6/7. So the probability of professor not meeting all racers is:

$$P(not \space meet \space all \space racers) = P(meet \space first \space racer)P(meet \space second \space racer)...P(meet \space seventh \space racer) = {(\frac{6}{7})}^7$$

So,

$$P(meet \space at \space least \space one \space racer) = 1 - P(not \space meet \space all \space racers) = 1 - {(\frac{6}{7})}^7 = 0.6600833229108862$$

In [1]:
pow(6,7)

279936

In [2]:
pow(7,7)

823543

In [3]:
823543 - 279936

543607

In [4]:
543607/823543

0.6600833229108862

In [5]:
P_Experiment1a = 0.6600833229108862

### b) By using combinatorics

In [6]:
class ProbDist(dict):
    """A Probability Distribution; an {outcome: probability} mapping."""
    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)
        # Make probabilities sum to 1.0; assert no negative probabilities
        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0

In [7]:
from fractions import Fraction
def p(event, space): 
    """The probability of an event, given a sample space of equiprobable outcomes. 
    event: a collection of outcomes, or a predicate that is true of outcomes in the event. 
    space: a set of outcomes or a probability distribution of {outcome: frequency} pairs."""
    # branch on the type of the first argument
    if is_predicate(event):
        # transform the mapping (untangible) 'event' into the collection (tangible) 'event'
        event = such_that(event, space)
        
    if isinstance(space, ProbDist):
        # if space is a dictionary of distinct probabilities, where each item does not count as the same amount
        return sum(space[k] for k in space if k in event)
        #return sum({space[k] for k in space & event})
    else:
        # space is not a dictionary but a collection, let's fall back to our original division
        return Fraction(len(event & space), len(space))

is_predicate = callable

def such_that(predicate, space): 
    """The outcomes in the sample pace for which the predicate is true.
    If space is a set, return a subset {outcome,...} with outcomes where predicate(element) is true;
    if space is a ProbDist, return a ProbDist {outcome: frequency,...} with outcomes where predicate(element) is true."""
    if isinstance(space, ProbDist):
        return ProbDist({o:space[o] for o in space if predicate(o)})
    else:
        return {o for o in space if predicate(o)}

We assume that the seven racers are: MV, CL, SP, GR, CS, LH and LN.

In [8]:
ProfessorAttend = ProbDist(Professor_Show1 = 1,
                           Professor_Show2 = 1,
                           Professor_Show3 = 1,
                           Professor_Show4 = 1,
                           Professor_Show5 = 1,
                           Professor_Show6 = 1,
                           Professor_Show7 = 1,)

In [9]:
ProfessorAttend

{'Professor_Show1': 0.14285714285714285,
 'Professor_Show2': 0.14285714285714285,
 'Professor_Show3': 0.14285714285714285,
 'Professor_Show4': 0.14285714285714285,
 'Professor_Show5': 0.14285714285714285,
 'Professor_Show6': 0.14285714285714285,
 'Professor_Show7': 0.14285714285714285}

In [10]:
MVAttend = ProbDist(MV_Show1 = 1,
                           MV_Show2 = 1,
                           MV_Show3 = 1,
                           MV_Show4 = 1,
                           MV_Show5 = 1,
                           MV_Show6 = 1,
                           MV_Show7 = 1,)
MVAttend

{'MV_Show1': 0.14285714285714285,
 'MV_Show2': 0.14285714285714285,
 'MV_Show3': 0.14285714285714285,
 'MV_Show4': 0.14285714285714285,
 'MV_Show5': 0.14285714285714285,
 'MV_Show6': 0.14285714285714285,
 'MV_Show7': 0.14285714285714285}

In [11]:
CLAttend = ProbDist(CL_Show1 = 1,
                           CL_Show2 = 1,
                           CL_Show3 = 1,
                           CL_Show4 = 1,
                           CL_Show5 = 1,
                           CL_Show6 = 1,
                           CL_Show7 = 1,)
CLAttend

{'CL_Show1': 0.14285714285714285,
 'CL_Show2': 0.14285714285714285,
 'CL_Show3': 0.14285714285714285,
 'CL_Show4': 0.14285714285714285,
 'CL_Show5': 0.14285714285714285,
 'CL_Show6': 0.14285714285714285,
 'CL_Show7': 0.14285714285714285}

In [12]:
SPAttend = ProbDist(SP_Show1 = 1,
                           SP_Show2 = 1,
                           SP_Show3 = 1,
                           SP_Show4 = 1,
                           SP_Show5 = 1,
                           SP_Show6 = 1,
                           SP_Show7 = 1,)
SPAttend

{'SP_Show1': 0.14285714285714285,
 'SP_Show2': 0.14285714285714285,
 'SP_Show3': 0.14285714285714285,
 'SP_Show4': 0.14285714285714285,
 'SP_Show5': 0.14285714285714285,
 'SP_Show6': 0.14285714285714285,
 'SP_Show7': 0.14285714285714285}

In [13]:
GRAttend = ProbDist(GR_Show1 = 1,
                           GR_Show2 = 1,
                           GR_Show3 = 1,
                           GR_Show4 = 1,
                           GR_Show5 = 1,
                           GR_Show6 = 1,
                           GR_Show7 = 1,)
GRAttend

{'GR_Show1': 0.14285714285714285,
 'GR_Show2': 0.14285714285714285,
 'GR_Show3': 0.14285714285714285,
 'GR_Show4': 0.14285714285714285,
 'GR_Show5': 0.14285714285714285,
 'GR_Show6': 0.14285714285714285,
 'GR_Show7': 0.14285714285714285}

In [14]:
CSAttend = ProbDist(CS_Show1 = 1,
                           CS_Show2 = 1,
                           CS_Show3 = 1,
                           CS_Show4 = 1,
                           CS_Show5 = 1,
                           CS_Show6 = 1,
                           CS_Show7 = 1,)
CSAttend

{'CS_Show1': 0.14285714285714285,
 'CS_Show2': 0.14285714285714285,
 'CS_Show3': 0.14285714285714285,
 'CS_Show4': 0.14285714285714285,
 'CS_Show5': 0.14285714285714285,
 'CS_Show6': 0.14285714285714285,
 'CS_Show7': 0.14285714285714285}

In [15]:
LHAttend = ProbDist(LH_Show1 = 1,
                           LH_Show2 = 1,
                           LH_Show3 = 1,
                           LH_Show4 = 1,
                           LH_Show5 = 1,
                           LH_Show6 = 1,
                           LH_Show7 = 1,)
LHAttend

{'LH_Show1': 0.14285714285714285,
 'LH_Show2': 0.14285714285714285,
 'LH_Show3': 0.14285714285714285,
 'LH_Show4': 0.14285714285714285,
 'LH_Show5': 0.14285714285714285,
 'LH_Show6': 0.14285714285714285,
 'LH_Show7': 0.14285714285714285}

In [16]:
LNAttend = ProbDist(LN_Show1 = 1,
                           LN_Show2 = 1,
                           LN_Show3 = 1,
                           LN_Show4 = 1,
                           LN_Show5 = 1,
                           LN_Show6 = 1,
                           LN_Show7 = 1,)
LNAttend

{'LN_Show1': 0.14285714285714285,
 'LN_Show2': 0.14285714285714285,
 'LN_Show3': 0.14285714285714285,
 'LN_Show4': 0.14285714285714285,
 'LN_Show5': 0.14285714285714285,
 'LN_Show6': 0.14285714285714285,
 'LN_Show7': 0.14285714285714285}

In [17]:
def joint8(A, B, C, D, E, F, G, H, sep=" "):
    """The joint distribution of 8 independent probability distributions. 
    Result is all entries of the form {a+sep+b+ ... for ...}"""
    return ProbDist({a + sep + b + sep + c + sep + d + sep + e + sep + f + sep + g + sep + h: A[a] * B[b]
                     * C[c] * D[d] * E[e] * F[f] * G[g] * H[h]
                    for a in A for b in B for c in C for d in D for e in E for f in F for g in G for h in H})

In [18]:
sample_space = joint8(ProfessorAttend, MVAttend, CLAttend, SPAttend, GRAttend, CSAttend, LHAttend, LNAttend)

In [19]:
sample_space

{'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show1': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show2': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show3': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show4': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show5': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show6': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show1 LN_Show7': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show2 LN_Show1': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 SP_Show1 GR_Show1 CS_Show1 LH_Show2 LN_Show2': 1.7346652554764097e-07,
 'Professor_Show1 MV_Show1 CL_Show1 S

In [20]:
def not_alone(outcome): return outcome.count(outcome.split(" ")[0].split("_")[1]) >= 2

In [21]:
P_Experiment1b = p(not_alone, sample_space)
P_Experiment1b

0.6600833228978629

In [22]:
import numpy as np

In [23]:
np.isclose(P_Experiment1a, P_Experiment1b)

True

The result is same as result from a)

To sum up, if there are 7 drivers the probability of professor meeting at least one driver is about **0.66**

## Experiment Two: one F1 driver goes to one or more of 7 shows for exactly 7 times and for 7 minutes each time

### a) By using theory and math or Monte Carlo simulations

So for example, the sole F1 driver could theoretically go to one show only, then leave, then come back, and repeat this seven times for a visit total of 49 minutes without going to any other show. 

One driver going to any of seven shows for a total of seven times. And we're looking to see if that driver is at the same show professor is  at. All possible combinations of one driver going to seven shows are equally likely, including one driver just going to one show seven times.

So they are many combinations like:
* (7) means driver goes to one show 7 times
* (6, 1) means driver goes to one show 6 times and one show 1 time
* (5, 1, 1) means driver goes to one show 5 times, one show 1 time and one show 1 tiem
* ...

Firstly, we compute the all the combinations:

In [24]:
from math import comb

In [25]:
countOfSituations = 0

In [26]:
countOfSituations = (7                    # go to one show for 7 times (7)
                    + 7 * 6 * 3           # (6, 1), (5, 2), (4, 3) 
                    + 7 * comb(6, 2) * 3  # (5, 1, 1), (3, 2, 2), (3, 3, 1)
                    + 7 * 6 * 5           # (4, 2, 1)
                    + 7 * comb(6, 3)      # (2, 2, 2, 1)
                    + 7 * 6 * comb(5, 2)  # (3, 2, 1, 1)
                    + 7 * comb(6, 4)      # (3, 1, 1, 1, 1)
                    + 7 * comb(6, 3)      # (2, 2, 2, 1)
                    + comb(7, 2) * comb(5, 3) # (2, 2, 1, 1, 1)
                    + 7 * comb(6, 5)          # (2, 1, 1, 1, 1, 1)
                    + 1)                      # (1, 1, 1, 1, 1, 1, 1)

In [27]:
countOfSituations

1716

We got that there are 1716 combinations totally.

After that we assume that professor attends a show and we can compute the conbinations of driver attending other 6 shows in order to get the probability of professor doesn't meet driver.

In [28]:
noProfessor =      (6                    # go to one show for 7 times (7)
                    + 6 * 5 * 3           # (6, 1), (5, 2), (4, 3) 
                    + 6 * comb(5, 2) * 3  # (5, 1, 1), (3, 2, 2), (3, 3, 1)
                    + 6 * 5 * 4           # (4, 2, 1)
                    + 6 * comb(5, 3)      # (2, 2, 2, 1)
                    + 6 * 5 * comb(4, 2)  # (3, 2, 1, 1)
                    + 6 * comb(5, 4)      # (3, 1, 1, 1, 1)
                    + 6 * comb(5, 3)      # (2, 2, 2, 1)
                    + comb(6, 2) * comb(4, 3) # (2, 2, 1, 1, 1)
                    + 6 * comb(5, 5)          # (2, 1, 1, 1, 1, 1)
                    + 0)

In [29]:
P_Experiment2a = 1 - (noProfessor/countOfSituations)
P_Experiment2a

0.5384615384615384

From this experiment we got that the probability of professor meeting dirver is **0.5384615384615384**

### b) By using combinatorics

We need to use python to generate all the prossible combinations. I generate strings to represent different combinations like:

* "7000000" represents driver goes to the first show 7 times
* "6010000" represents driver goes to the first show 6 times and the third show 1 time
* ...

In [30]:
import random
random.seed(10)

I use random number to generate random string where the sum of elements is 7. And I add these strings to a set so that I can make sure all the strings are unique.

In [31]:
sampleSet = set()

In [32]:
def getAPossibleSituation(): # generate a combination(string) randomly
    times = 7
    situation = ""
    for i in range(6):
        currentTimes = random.randint(0, times)
        situation += str(currentTimes)
        times -= currentTimes
    situation += str(times)
    return situation

In [33]:
getAPossibleSituation()

'0610000'

We generate 3000000 random strings. This is a huge number so that we can make sure we cover all the combinations. As a result, the code below takes a few time.

In [34]:
for i in range(3000000): 
    sampleSet.add(getAPossibleSituation())

In [35]:
len(sampleSet)

1716

The length of sampleSet is 1716. So we exactly covered all possible combinations. 

To accelerate the computing, I modified the not_alone and p function to make them quicker.

For the not_alone function, I assume that professor attends the first show. So I just neet to check the first element of the strings, if it is 0 they don't meet, if it is not 0, they meet.

In [36]:
def not_alone_2(outcome):
    if int(outcome[0]) != 0:
        return True

In [37]:
def p_2(not_alone_2, sampleSet):
    count = 0
    for i, situation in enumerate(sampleSet):
        if (not_alone_2(situation)):
            count += 1
    return count/len(sampleSet)

In [38]:
P_Experiment2b = p_2(not_alone_2, sampleSet)
P_Experiment2b

0.5384615384615384

In [39]:
np.isclose(P_Experiment2a, P_Experiment2b)

True

The result is same as result from a)

To sum up, if there are 7 drivers the probability of they meet each other is **0.5384615384615384**

## Conclusioin

In [40]:
P_Experiment1b > P_Experiment2b

True

The first experiment gives professor the higher probabilities of meeting a F1 driver